java-manta-client-unshaded/src/main/java/com/joyent/manta/client/PruneEmptyParentDirectoryStrategy.java

```diff
@@ -0,0 +1,149 @@
+/*
+ * Copyright (c) 2017, Joyent, Inc. All rights reserved.
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+package com.joyent.manta.client;
+
+import com.joyent.manta.exception.MantaClientHttpResponseException;
+import com.joyent.manta.http.MantaHttpHeaders;
+import com.joyent.manta.util.MantaUtils;
+import org.apache.commons.lang3.ArrayUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+
+/**
+ * Utility class for recursive directory removal strategies.
+ *
+ * @author <a href="https://github.com/douglasAtJoyent">Douglas Anderson</a>
+ * @since 3.2.4
+ */
+final class PruneEmptyParentDirectoryStrategy {
+
+    private static final Logger LOG = LoggerFactory.getLogger(PruneEmptyParentDirectoryStrategy.class);
+
+    /**
+     * Pass this if you want to prune all parent directories.
+     */
+    public static final Integer PRUNE_ALL_PARENTS = -1;
+
+    private PruneEmptyParentDirectoryStrategy() {
+    }
+
+    /**
+     * Method that will delete the specified path. If the delete of the path is
+     * successful, then it will progressively go from right (deepest) to left
+     * in the path attempting to delete all parent directories until a non-empty
+     * directory is encountered or until a directory is missing.
+     *
+     * @param client a reference to the manta client instance for the operations
+     * @param path file or directory object to delete
+     * @param limit limit of directories to delete
+     * @throws IOException thrown when there is a problem deleting from Manta
+     */
+    static void pruneParentDirectories(final MantaClient client,
+            final MantaHttpHeaders headers,
+            final String path,
+            final int limit) throws IOException {
```

Comment:
Would it make sense to call this 'depthLimit' or 'maxDepth'? I think the name of this variable should imply that the limit is on how far up the hierarchy we search for pruning opportunities (and not just on how many prunes we attempt).

Reply:
I opted for `maxDepth`.